src/analysis/ppx_expand.ml

```diff
@@ -0,0 +1,155 @@
+        attribute.attr_name.txt = "deriving"
+        && check_at_pos pos attribute.attr_loc)
+      attrs
+  in
+  match found_attr with
+  | Some attribute -> Some attribute.attr_loc
+  | None -> None
+
+let check_structures pos (item : Parsetree.structure_item_desc) =
+  match item with
+  | Pstr_type (_, ty) ->
+      List.find_map
+        (fun (t : Parsetree.type_declaration) ->
+          check_deriving_attr pos t.ptype_attributes)
+        ty
+  | Pstr_exception tc -> check_deriving_attr pos tc.ptyexn_attributes
+  | Pstr_modtype mt -> check_deriving_attr pos mt.pmtd_attributes
+  | Pstr_extension (_, attrs) -> check_deriving_attr pos attrs
+  | _ -> None
+
+let check_signatures pos (item : Parsetree.signature_item_desc) =
+  match item with
+  | Psig_type (_, ty) ->
+      List.find_map
+        (fun (t : Parsetree.type_declaration) ->
+          check_deriving_attr pos t.ptype_attributes)
+        ty
+  | Psig_exception tc -> check_deriving_attr pos tc.ptyexn_attributes
+  | Psig_modtype mt -> check_deriving_attr pos mt.pmtd_attributes
+  | Psig_extension (_, attrs) -> check_deriving_attr pos attrs
+  | _ -> None
+
+let check_extension ~parsetree ~pos =
+  let kind = ref None in
+  let expr (self : Ast_iterator.iterator) (expr : Parsetree.expression) =
+    match check_extension_node pos expr with
+    | Some ext_loc -> kind := Some (Expr expr, ext_loc)
+    | None -> Ast_iterator.default_iterator.expr self expr
+  in
+  let signature_item (self : Ast_iterator.iterator)
+      (original_sg : Parsetree.signature_item) =
+    match check_signatures pos original_sg.psig_desc with
+    | Some attr_loc -> kind := Some (Sig_item original_sg, attr_loc)
+    | None -> Ast_iterator.default_iterator.signature_item self original_sg
+  in
+  let structure_item (self : Ast_iterator.iterator)
+      (original_str : Parsetree.structure_item) =
+    match check_structures pos original_str.pstr_desc with
+    | Some attr_loc -> kind := Some (Str_item original_str, attr_loc)
+    | None -> Ast_iterator.default_iterator.structure_item self original_str
+  in
+  let iterator =
+    { Ast_iterator.default_iterator with signature_item; structure_item; expr }
+  in
+  let () =
+    match parsetree with
+    | `Interface si -> iterator.signature iterator si
+    | `Implementation str -> iterator.structure iterator str
+  in
+  !kind
+
+let get_ppxed_source ~ppxed_parsetree ~pos ppx_kind_with_attr :
+    Query_protocol.ppxed_source =
+  let expression = ref [] in
+  let signature = ref [] in
+  let structure = ref [] in
+  let () =
+    match ppx_kind_with_attr with
+    | Expr original_expr, _ -> (
+        let expr (self : Ast_iterator.iterator)
+            (new_expr : Parsetree.expression) =
+          match
+            original_expr.pexp_loc.loc_start.pos_cnum
+            <= new_expr.pexp_loc.loc_start.pos_cnum
+            && new_expr.pexp_loc.loc_end.pos_cnum
+               <= original_expr.pexp_loc.loc_end.pos_cnum
```

Comment:
Are you sure that works with multi-line expressions ?
You probably want to use `Location_aux.included`

Reply:
This is a very cool function.